File: demo/services/worker.py
"""Background job worker: progress bars, deep stack traces, and a stress mode.

This is the service to pin when you want to see composemux handle output that
is not plain lines - carriage returns redrawing a bar in place, and tracebacks
tall enough to scroll.
"""

import json
import os
import random
import sys
import time
import traceback

sys.path.insert(0, os.path.dirname(__file__))
from common import (  # noqa: E402
    CYAN, GREEN, GREY, JOBS, MAGENTA, RED, RESET, YELLOW,
    deep_traceback, flag, install_clean_exit, jitter, log, progress,
)

IDLE_CHATTER = [
    ("DEBUG", "poll: queue empty, backing off"),
    ("DEBUG", "heartbeat sent to coordinator"),
    ("INFO", "lease renewed for partition {p}"),
    ("DEBUG", "compacted {n} tombstones"),
]


def take_job():
    """Claim the oldest job file, or None. Renaming is the claim."""
    try:
        names = sorted(n for n in os.listdir(JOBS) if n.endswith(".json"))
    except FileNotFoundError:
        return None
    for name in names:
        path = os.path.join(JOBS, name)
        claimed = path + ".taken"
        try:
            os.rename(path, claimed)
        except OSError:
            continue  # another replica got there first
        try:
            with open(claimed) as fh:
                return json.load(fh)
        finally:
            os.unlink(claimed)
    return None


def run_job(job):
    kind, jid, size = job["kind"], job["id"], job["size"]
    log("INFO", f"{MAGENTA}job {jid}{RESET} start kind={CYAN}{kind}{RESET} size={size}")

    steps = max(10, min(40, size // 200))
    progress(f"job {jid} {kind}", steps=steps, delay=0.08)

    # Roughly one job in six falls over, which is what makes the pane worth
    # scrolling back through.
    if random.random() < 0.17:
        try:
            deep_traceback()
        except RuntimeError:
            log("ERROR", f"{RED}job {jid} failed after {steps} steps{RESET}")
            for line in traceback.format_exc().rstrip().splitlines():
                log("ERROR", f"{GREY}{line}{RESET}")
            log("WARN", f"job {jid} scheduled for retry in 30s")
            return

    log("INFO", f"{MAGENTA}job {jid}{RESET} {GREEN}done{RESET} in {steps * 0.05:.1f}s")


def stress_burst():
    """High-volume output, for watching a scrolled-up pane hold its position."""
    log("WARN", f"{YELLOW}stress mode engaged - flooding stdout{RESET}")
    n = 0
    while flag("stress"):
        n += 1
        shard = n % 8
        log(
            "DEBUG",
            f"{GREY}batch={n:06d} shard={shard} rows={random.randint(50, 900)} "
            f"lag={random.randint(0, 400)}ms offset={random.randint(10**6, 10**7)}{RESET}",
        )
        time.sleep(0.01)
    log("INFO", f"{GREEN}stress mode released after {n} lines{RESET}")


def main():
    install_clean_exit()
    replica = os.environ.get("REPLICA", "?")
    log("INFO", f"worker starting (container {os.uname().nodename})")
    log("INFO", f"{GREEN}ready{RESET} - polling {JOBS}")
    tick = 0

    while True:
        if flag("stress"):
            stress_burst()
            continue

        job = take_job()
        if job:
            run_job(job)
            continue

        tick += 1
        if tick % 4 == 0:
            level, text = random.choice(IDLE_CHATTER)
            log(level, text.format(p=random.randint(0, 7), n=random.randint(1, 400)))
        time.sleep(jitter(0.8))


if __name__ == "__main__":
    main()

File: demo/services/worker.ts
/**
 * Background job worker: progress bars, deep stack traces, and a stress mode.
 *
 * This is the service to pin when you want to see composemux handle output that
 * is not plain lines - carriage returns redrawing a bar in place, and stack
 * traces tall enough to scroll.
 */

import { readdirSync, readFileSync, renameSync, unlinkSync } from "node:fs";
import { hostname } from "node:os";

import {
  CYAN, GREEN, GREY, JOBS, MAGENTA, RED, RESET, YELLOW,
  choice, deepThrow, flag, formatError, installCleanExit, jitter, log, progress, randint, sleep,
  type Level,
} from "./common.ts";

type Job = { id: string; kind: string; size: number };

const IDLE_CHATTER: ReadonlyArray<readonly [Level, string]> = [
  ["DEBUG", "poll: queue empty, backing off"],
  ["DEBUG", "heartbeat sent to coordinator"],
  ["INFO", "lease renewed for partition {p}"],
  ["DEBUG", "compacted {n} tombstones"],
];

/** Claim the oldest job file, or null. Renaming is the claim. */
function takeJob(): Job | null {
  let names: string[];
  try {
    names = readdirSync(JOBS).filter((n) => n.endsWith(".json")).sort();
  } catch {
    return null;
  }
  for (const name of names) {
    const path = `${JOBS}/${name}`;
    const claimed = `${path}.taken`;
    try {
      renameSync(path, claimed);
    } catch {
      continue; // another replica got there first
    }
    try {
      return JSON.parse(readFileSync(claimed, "utf8")) as Job;
    } finally {
      unlinkSync(claimed);
    }
  }
  return null;
}

async function runJob(job: Job): Promise<void> {
  const { kind, id, size } = job;
  log("INFO", `${MAGENTA}job ${id}${RESET} start kind=${CYAN}${kind}${RESET} size=${size}`);

  const steps = Math.max(10, Math.min(40, Math.floor(size / 200)));
  await progress(`job ${id} ${kind}`, steps, 0.08);

  // Roughly one job in six falls over, which is what makes the pane worth
  // scrolling back through.
  if (Math.random() < 0.17) {
    try {
      deepThrow();
    } catch (err) {
      log("ERROR", `${RED}job ${id} failed after ${steps} steps${RESET}`);
      for (const line of formatError(err)) log("ERROR", `${GREY}${line}${RESET}`);
      log("WARN", `job ${id} scheduled for retry in 30s`);
      return;
    }
  }

  log("INFO", `${MAGENTA}job ${id}${RESET} ${GREEN}done${RESET} in ${(steps * 0.05).toFixed(1)}s`);
}

/** High-volume output, for watching a scrolled-up pane hold its position. */
async function stressBurst(): Promise<void> {
  log("WARN", `${YELLOW}stress mode engaged - flooding stdout${RESET}`);
  let n = 0;
  while (flag("stress")) {
    n += 1;
    log(
      "DEBUG",
      `${GREY}batch=${String(n).padStart(6, "0")} shard=${n % 8} rows=${randint(50, 900)} ` +
        `lag=${randint(0, 400)}ms offset=${randint(10 ** 6, 10 ** 7)}${RESET}`,
    );
    await sleep(0.01);
  }
  log("INFO", `${GREEN}stress mode released after ${n} lines${RESET}`);
}

async function main(): Promise<void> {
  installCleanExit();
  log("INFO", `worker starting (container ${hostname()})`);
  log("INFO", `${GREEN}ready${RESET} - polling ${JOBS}`);
  let tick = 0;

  for (;;) {
    if (flag("stress")) {
      await stressBurst();
      continue;
    }

    const job = takeJob();
    if (job) {
      await runJob(job);
      continue;
    }

    tick += 1;
    if (tick % 4 === 0) {
      const [level, text] = choice(IDLE_CHATTER);
      log(level, text.replace("{p}", String(randint(0, 7))).replace("{n}", String(randint(1, 400))));
    }
    await sleep(jitter(0.8));
  }
}

await main();
